catalyst-gateway/bin/src/service/api/cardano/cip36/filter.rs

```diff
@@ -1,643 +1,440 @@
+                    Err(err) => {
+                        anyhow::bail!("Corrupt invalid registration {err}");
+                    },
+                };
+                let payment_address = Cip19Address::try_from(row.payment_address).ok();
+                let vote_pub_key = Ed25519HexEncodedPublicKey::try_from(row.vote_key).ok();
+                let stake_pub_key =
+                    Ed25519HexEncodedPublicKey::try_from(row.stake_public_key.clone()).ok();
+                let report: serde_json::Value =
+                    serde_json::from_str(&row.problem_report).map_err(|e| {
+                        anyhow::anyhow!("Invalid CIP36 registration problem report should me JSON encodable, err: {e}")
+                    })?;
+
+                regs.push(Cip36Details {
+                    slot_no,
+                    stake_pub_key,
+                    vote_pub_key,
+                    nonce: Some(nonce.into()),
+                    raw_nonce: Some(raw_nonce.into()),
+                    txn_index: Some(row.txn_index.into()),
+                    payment_address,
+                    is_payable: row.is_payable.into(),
+                    cip15: (!row.cip36).into(),
+                    report: Some(report.into()),
+                });
+                Ok(regs)
+            }
+        })
+        .await
+}
 
-    while let Some(row) = registrations_iter.next().await {
-        let row = row?;
+/// Get all valid cip36 registrations.
+async fn get_all_valid_registrations(
+    session: &CassandraSession, slot_no: SlotNo,
+) -> Result<Vec<Cip36Details>, anyhow::Error> {
+    let regs_stream =
+        GetAllRegistrationsQuery::execute(session, GetAllRegistrationsParams::new(slot_no.into()))
+            .await?;
 
-        let nonce = if let Some(nonce) = row.nonce.into_parts().1.to_u64_digits().first() {
-            *nonce
-        } else {
-            continue;
+    regs_stream.map_err(Into::<anyhow::Error>::into).try_fold(Vec::new(), |mut regs, row|
+    async move {
+        let Ok(nonce) = u64::try_from(row.nonce) else {
+            anyhow::bail!("Corrupt valid registration, cannot decode nonce");
         };
-
-        let slot_no = if let Some(slot_no) = row.slot_no.into_parts().1.to_u64_digits().first() {
-            *slot_no
-        } else {
-            continue;
+        let Ok(raw_nonce) = u64::try_from(row.raw_nonce) else {
+            anyhow::bail!("Corrupt valid registration, cannot decode raw_nonce");
         };
 
-        let slot_no = match SlotNo::try_from(slot_no) {
+        let slot_no = match SlotNo::try_from(u64::from(row.slot_no)) {
             Ok(slot_no) => slot_no,
             Err(err) => {
-                error!("Corrupt valid registration {:?}", err);
-                // This should NOT happen, valid registrations should be infallible.
-                // If it happens, there is an indexing issue.
-                continue;
+                anyhow::bail!("Corrupt invalid registration {err}");
```

Comment:
Why are we bailing, and not just ignoring a registration which may not be the latest?  This prevents us finding valid registrations because something was wrong.